sdk/src/nonce_state.rs

```diff
@@ -0,0 +1,545 @@
+use serde_derive::{Deserialize, Serialize};
+use crate::{
+    account::{Account, KeyedAccount},
+    account_utils::State,
+    hash::Hash,
+    instruction::InstructionError,
+    nonce_program::id,
+    nonce_instruction::{NonceError},
+    pubkey::Pubkey,
+    sysvar::recent_blockhashes::RecentBlockhashes,
+    sysvar::rent::Rent,
+};
+use std::collections::HashSet;
+
+#[derive(Debug, Default, Serialize, Deserialize, PartialEq, Clone, Copy)]
+pub struct Meta {
+    rent_exempt_lamports: u64,
+}
+
+impl Meta {
+    pub fn new(rent_exempt_lamports: u64) -> Self {
+        Self {
+            rent_exempt_lamports,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone, Copy)]
+pub enum NonceState {
+    Uninitialized,
+    Initialized(Meta, Hash),
+}
+
+impl Default for NonceState {
+    fn default() -> Self {
+        Self::Uninitialized
+    }
+}
+
+impl NonceState {
+    pub fn size() -> usize {
+        bincode::serialized_size(
+            &NonceState::Initialized(
+                Meta::default(),
+                Hash::default(),
+        )).unwrap() as usize
+    }
+}
+
+pub trait NonceAccount {
+    fn nonce(
+        &mut self,
+        recent_blockhashes: &RecentBlockhashes,
+        rent: &Rent,
+    ) -> Result<(), InstructionError>;
+    fn set_nonce(
+        &mut self,
+        recent_blockhashes: &RecentBlockhashes,
+        meta: &Meta,
+    ) -> Result<(), InstructionError>;
+    fn withdraw(
+        &mut self,
+        lamports: u64,
+        to: &mut KeyedAccount,
+        recent_blockhashes: &RecentBlockhashes,
+        signers: &HashSet<Pubkey>,
+    ) -> Result<(), InstructionError>;
+}
+
+impl<'a> NonceAccount for KeyedAccount<'a> {
+    fn set_nonce(
+        &mut self,
+        recent_blockhashes: &RecentBlockhashes,
+        meta: &Meta,
+    ) -> Result<(), InstructionError> {
+        if recent_blockhashes.is_empty() {
```

Comment:
I think this check could get moved to the beginning of `fn nonce(..)` and then `set_nonce` can be passed a blockhash directly

Reply:
I'm actually leaning towards dumping `set_nonce` altogether.  The code reuse that inspired it no longer exists.